### PR TITLE
service: Handle single stack service errors when dual-stack flag enabled

### PR DESCRIPTION
**Pull request by smarterclayton**
If the dual-stack flag is enabled and the cluster is single stack IPv6,
the allocator logic for service clusterIP does not properly handle rejecting
a request for an IPv4 family. Return a 422 Invalid on the ipFamily field
when the dual stack flag is on (as it would when it hits beta) and the
cluster is configured for single-stack IPv6. Without this check, the allocator
would allocate an IPv6 cluster IP to a service requesting an IPv4 family.

Found by observation when testing an IPv6 single stack cluster with the
feature gate enabled. 

/kind bug

```release-note
When the DualStack flag is enabled, a single stack IPv6 cluster will correctly reject services that request a cluster IP with an IPv4 family (since the cluster has no IPv4 service CIDR configured).
```

```docs
```

### CODE REVIEW

**Comment by smarterclayton** [comment]
/assign @dcbw 

as we discovered today

**Comment by smarterclayton** [comment]
Hrm, I think this needs to be in validation (like other strategies that depend on config).  I’ll rework this a bit tonight

/hold

File: pkg/registry/core/service/storage/rest.go [hunk 1]
```diff
@@ -666,6 +666,21 @@ func allocateHealthCheckNodePort(service *api.Service, nodePortOp *portallocator
 func initClusterIP(service *api.Service, allocator ipallocator.Interface) (bool, error) {
 	switch {
 	case service.Spec.ClusterIP == "":
+		// ensure the IP we will allocate is valid within the configured allocator for the cluster
```
Comment (smarterclayton) [inline pkg/registry/core/service/storage/rest.go]: This needs to be in validation and the strategy needs to be parameterized with the correct settings (single stack, dual stack).

**Comment by smarterclayton** [comment]
I think #86895 is a more correct (although more invasive) fix, which properly moves defaulting the ipfamily value from the storage layer into the strategy, and moves checking membership in the family into the validation layer invoked by the strategy.  @liggitt if you have any strong opinions here (generally strategies hold validation and prep logic, not registries)

**Comment by liggitt** [comment]
I do prefer the strategy doing this over the storage

**Comment by aojea** [comment]
the ipv6 test failure is unrelated and is my fault :man_facepalming:

**Comment by smarterclayton** [comment]
Closed in favor of the alternate (which fixes the "can't update a service on a single stack ipv6 cluster with old client and dual stack flag enabled".

**Pull request closed without merging**